File: Strand/Screens/CoachLauncherSheet.swift
import SwiftUI
import StrandDesign

/// The compact Coach launcher opened from the optional Today card (#1862).
///
/// Coach is otherwise reachable only through More/Insights, which makes it easy to miss and means
/// leaving Today to try it. This is the shortcut — and deliberately ONLY a shortcut.
///
/// It owns no send, stream, error or consent surface of its own. Picking a suggestion or submitting the
/// composer hands the question to `AICoachEngine.pendingPrompt` and routes to `CoachView`, which already
/// has all of that. A second chat UI would drift from the first, and the issue explicitly defers the
/// persistent-workspace redesign (threads, retention, backup policy) to a follow-up.
///
/// NO PROVIDER REQUEST IS MADE BY OPENING THIS. Everything shown is local: `isConfigured` reads the
/// stored key, and the prompts are static copy. The first network call still happens where it always
/// did — inside `AICoachEngine.send`, after an explicit user action.
struct CoachLauncherSheet: View {
    @EnvironmentObject var coach: AICoachEngine
    @EnvironmentObject var router: NavRouter
    @Environment(\.dismiss) private var dismiss

    @State private var draft = ""

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 16) {
                    if coach.isConfigured {
                        configured
                    } else {
                        unconfigured
                    }
                }
                .padding(16)
                .frame(maxWidth: .infinity, alignment: .leading)
            }
            .background(StrandPalette.surfaceBase.ignoresSafeArea())
            .navigationTitle(Text("Coach"))
            #if os(iOS)
            .navigationBarTitleDisplayMode(.inline)
            #endif
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                }
            }
        }
        #if os(iOS)
        .presentationDetents([.medium, .large])
        #endif
    }

    // MARK: Configured — suggestions + a compact composer

    @ViewBuilder
    private var configured: some View {
        Text("Ask about your charge, effort, rest and workouts, grounded in your own numbers.")
            .font(StrandFont.footnote)
            .foregroundStyle(StrandPalette.textTertiary)

        // Deliberately no "Try asking" heading: the line above already frames the list, and a new
        // literal would be the only string in this card needing translation into ten locales. The whole
        // launcher reuses copy the Coach screen already ships.

        ForEach(CoachPrompts.suggestions, id: \.self) { prompt in
            Button { hand(off: prompt) } label: {
                Text(prompt)
                    .font(StrandFont.caption)
                    .foregroundStyle(StrandPalette.textPrimary)
                    .padding(.horizontal, 12).padding(.vertical, 8)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .background(FrostedCardSurface(cornerRadius: NoopMetrics.cardRadius))
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(Text("Suggested prompt: \(prompt)"))
        }

        HStack(spacing: 8) {
            TextField("Ask your coach…", text: $draft, axis: .vertical)
                .textFieldStyle(.plain)
                .lineLimit(1...3)
                .padding(.horizontal, 12).padding(.vertical, 9)
                .background(FrostedCardSurface(cornerRadius: NoopMetrics.cardRadius))
                .onSubmit { submitDraft() }
            Button {
                submitDraft()
            } label: {
                Image(systemName: "arrow.up.circle.fill")
                    .font(.system(size: 26, weight: .semibold))
                    .foregroundStyle(StrandPalette.accent)
            }
            .buttonStyle(.plain)
            .disabled(trimmedDraft.isEmpty)
            .accessibilityLabel(Text("Send"))
        }
        .padding(.top, 4)
    }

    // MARK: Not configured — explain the opt-in and route to the existing setup

    @ViewBuilder
    private var unconfigured: some View {
        // The SAME explanation the Coach screen shows, so the bring-your-own-key model is described
        // once. The button routes to that screen, which stays the only place a key is entered.
        Text("Coach uses your own API key. Pick a provider, paste a key, and choose a model. Your key is stored securely in the Keychain and never leaves \(Platform.deviceNounPhrase) except as the request you make.")
            .font(StrandFont.footnote)
            .foregroundStyle(StrandPalette.textTertiary)

        Button {
            dismiss()
            router.openCoach()
        } label: {
            Text("Connect a provider")
                .font(StrandFont.caption)
                .foregroundStyle(StrandPalette.textPrimary)
                .padding(.horizontal, 14).padding(.vertical, 10)
                .frame(maxWidth: .infinity)
                .background(FrostedCardSurface(cornerRadius: NoopMetrics.cardRadius))
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    // MARK: Handoff

    private var trimmedDraft: String {
        draft.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func submitDraft() {
        let text = trimmedDraft
        guard !text.isEmpty else { return }
        hand(off: text)
    }

    /// Dismiss, park the question, and open Coach. Deliberately NOT `coach.send` from here: the launcher
    /// never performs a provider request, so a user who opens this sheet and changes their mind has cost
    /// nothing and sent nothing.
    private func hand(off prompt: String) {
        coach.pendingPrompt = prompt
        draft = ""
        dismiss()
        router.openCoach()
    }
}

File: android/app/src/main/java/com/noop/ui/CoachLauncherSheet.kt
package com.noop.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.ui.draw.clip
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.material3.Text
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.noop.R

/**
 * The compact Coach launcher opened from the optional Today card (#1862).
 *
 * Coach is otherwise reachable only through More, which makes it easy to miss and means leaving Today to
 * try it. This is the shortcut — and deliberately ONLY a shortcut.
 *
 * It owns no send, stream, error or consent surface of its own. Picking a suggestion hands the question
 * to [onPick], which routes to the Coach screen; that screen already has all of those. A second chat UI
 * would drift from the first, and #1862 explicitly defers the persistent-workspace redesign (threads,
 * retention, backup policy) to a follow-up.
 *
 * NO PROVIDER REQUEST IS MADE BY SHOWING THIS. [isConfigured] is a local key read and the prompts are
 * static copy; the first network call still happens where it always did, after an explicit send.
 *
 * Presentational by construction — every input is injected — so it can be exercised without a provider,
 * a key, or a ViewModel. Swift twin: `CoachLauncherSheet`.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CoachLauncherSheet(
    isConfigured: Boolean,
    onPick: (String) -> Unit,
    onSetup: () -> Unit,
    onDismiss: () -> Unit,
) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = Palette.surfaceRaised,
    ) {
        Column(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 8.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Overline(uiString(R.string.nav_coach))
            // Frames BOTH branches, so the sheet says what Coach is before it either offers questions or
            // asks for a provider. Using it once also avoids the duplicate that an unconfigured-only
            // explainer produced, where "Connect a provider" read as both the description and the button.
            Text(
                uiString(R.string.l10n_coach_screen_ask_anything_about_your_recent_recovery_e6c287ca),
                style = NoopType.caption,
                color = Palette.textTertiary,
            )
            if (isConfigured) {
                CoachPrompts.SUGGESTIONS.forEach { prompt ->
                    PromptRow(prompt = prompt, onClick = { onPick(prompt) })
                }
            } else {
                // The Coach screen stays the only place a key is entered; this just routes there.
                PromptRow(
                    prompt = uiString(R.string.l10n_coach_screen_connect_a_provider_6967f288),
                    onClick = onSetup,
                )
            }
        }
    }
}

@Composable
private fun PromptRow(prompt: String, onClick: () -> Unit) {
    val shape = RoundedCornerShape(12.dp)
    Text(
        prompt,
        style = NoopType.caption,
        color = Palette.textPrimary,
        modifier = Modifier
            .fillMaxWidth()
            .clip(shape)
            .background(Palette.surfaceInset)
            .border(1.dp, Palette.hairline, shape)
            .clickable { onClick() }
            .padding(horizontal = 12.dp, vertical = 10.dp)
            .semantics { contentDescription = prompt },
    )
}
